In [51]:
import pandas as pd
import matplotlib.pyplot as plt

In [52]:
sldb = pd.read_csv('../data/synergy_logistics_database.csv')

## 1 Rutas de importación y exportación
Las rutas se definirán como una combinación de:
* Origen, destino y medio de transporte

In [53]:
exportaciones = sldb[sldb['direction']=='Exports']
importaciones = sldb[sldb['direction']=='Imports']
ganancias_totales = sldb['total_value'].sum()


In [55]:
rutas_exp = exportaciones.groupby(['origin','destination','transport_mode'])
suma_exp = rutas_exp.sum()['total_value']
rutas_exp = rutas_exp['total_value'].describe()
rutas_exp['suma total'] = suma_exp
rutas_exp = rutas_exp.reset_index()

In [56]:
top_rutas_exp = rutas_exp.sort_values(by = 'count', ascending=False).head(10)
top_rutas_exp

origin  destination transport_mode  count          mean  \
134  South Korea      Vietnam            Sea  497.0  1.383704e+07   
154          USA  Netherlands            Sea  436.0  2.367401e+06   
108  Netherlands      Belgium           Road  374.0  8.636615e+06   
34         China       Mexico            Air  330.0  3.712121e+07   
86         Japan       Brazil            Sea  306.0  1.100704e+07   
57       Germany       France           Road  299.0  9.706401e+06   
131  South Korea        Japan            Sea  279.0  1.646595e+07   
3      Australia    Singapore            Sea  273.0  1.805861e+06   
25        Canada       Mexico           Rail  261.0  3.237548e+07   
38         China        Spain            Air  250.0  7.448000e+06   

              std        min         25%         50%         75%          max  \
134  1.071060e+07        0.0   6000000.0  11000000.0  18000000.0   65000000.0   
154  3.425529e+06        0.0      1000.0      4000.0   4000000.0   23000000.0   
108  1.787992e+07        0.0      1000.0   3000000.0   7000000.0  229000000.0   
34   2.582412e+07        0.0  19000000.0  35000000.0  52000000.0  140000000.0   
86   1.115415e+07        0.0   1000000.0   8000000.0  17000000.0   52000000.0   
57   1.577356e+07        0.0      2000.0   3000000.0  10000000.0   88000000.0   
131  8.336327e+06  2000000.0  10000000.0  15000000.0  20500000.0   45000000.0   
3    1.607325e+06        0.0   1000000.0   1000000.0   2000000.0   12000000.0   
25   2.433130e+07        0.0  11000000.0  31000000.0  49000000.0  103000000.0   
38   9.225385e+06        0.0   2000000.0   4000000.0   8000000.0   54000000.0   

      suma total  
134   6877007000  
154   1032187000  
108   3230094000  
34   12250000000  
86    3368155000  
57    2902214000  
131   4594000000  
3      493000000  
25    8450000000  
38    1862000000

In [ ]:
rutas_imp = importaciones.groupby(['origin','destination','transport_mode'])
suma_imp = rutas_imp.sum()['total_value']
rutas_imp = rutas_imp['total_value'].describe()
rutas_imp['suma total']=suma_imp
top_rutas_imp = rutas_imp.sort_values(by='count',ascending=False).head(10)
top_rutas_imp

count          mean  \
origin    destination          transport_mode                        
Singapore Thailand             Sea             273.0  1.471429e+07   
Germany   China                Sea             233.0  5.699571e+06   
China     Japan                Air             210.0  1.541429e+07   
Japan     Mexico               Sea             206.0  1.901942e+07   
Malaysia  Thailand             Rail            195.0  1.785641e+07   
China     Thailand             Road            145.0  1.921379e+07   
Spain     Germany              Road            142.0  1.439437e+07   
China     United Arab Emirates Sea             114.0  2.944737e+07   
Brazil    China                Sea             113.0  3.654867e+06   
USA       Thailand             Sea             109.0  1.438532e+07   

                                                        std        min  \
origin    destination          transport_mode                            
Singapore Thailand             Sea             1.216850e+07  1000000.0   
Germany   China                Sea             4.801656e+06        0.0   
China     Japan                Air             1.151013e+07  1000000.0   
Japan     Mexico               Sea             1.764251e+07        0.0   
Malaysia  Thailand             Rail            1.545128e+07  1000000.0   
China     Thailand             Road            1.652741e+07  1000000.0   
Spain     Germany              Road            1.399035e+07        0.0   
China     United Arab Emirates Sea             2.043123e+07  1000000.0   
Brazil    China                Sea             2.701752e+06        0.0   
USA       Thailand             Sea             9.922765e+06  1000000.0   

                                                     25%         50%  \
origin    destination          transport_mode                          
Singapore Thailand             Sea             4000000.0  14000000.0   
Germany   China                Sea             2000000.0   5000000.0   
China     Japan                Air             3000000.0  14500000.0   
Japan     Mexico               Sea             4000000.0  11000000.0   
Malaysia  Thailand             Rail            3000000.0  17000000.0   
China     Thailand             Road            5000000.0  15000000.0   
Spain     Germany              Road            3000000.0   5500000.0   
China     United Arab Emirates Sea             6000000.0  34500000.0   
Brazil    China                Sea             1000000.0   4000000.0   
USA       Thailand             Sea             3000000.0  16000000.0   

                                                      75%         max  \
origin    destination          transport_mode                           
Singapore Thailand             Sea             20000000.0  69000000.0   
Germany   China                Sea              8000000.0  23000000.0   
China     Japan                Air             23000000.0  46000000.0   
Japan     Mexico               Sea             36000000.0  61000000.0   
Malaysia  Thailand             Rail            25000000.0  69000000.0   
China     Thailand             Road            27000000.0  78000000.0   
Spain     Germany              Road            27750000.0  45000000.0   
China     United Arab Emirates Sea             46000000.0  85000000.0   
Brazil    China                Sea              5000000.0  12000000.0   
USA       Thailand             Sea             21000000.0  41000000.0   

                                               suma total  
origin    destination          transport_mode              
Singapore Thailand             Sea             4017000000  
Germany   China                Sea             1328000000  
China     Japan                Air             3237000000  
Japan     Mexico               Sea             3918000000  
Malaysia  Thailand             Rail            3482000000  
China     Thailand             Road            2786000000  
Spain     Germany              Road            2044000000  
China     United Arab Emirates Sea 

In [74]:
def ganancias_top(df,top):
    suma_total = df['suma total'].sum()
    top_rutas = df.sort_values(by='suma total', ascending = False).head(top)
    suma_top = top_rutas['suma total'].sum()

    repeticiones = top_rutas['count'].sum()
    porcentaje = (suma_top/suma_total)*1000
    porcentaje = int(porcentaje)/10

    print(f'Las {top} rutas mas demandadas aportan el {porcentaje}% de las ganancias.\nTotal de servicios realizados fue de: {repeticiones} servicios\n')

print("Rutas de exportación")
ganancias_top(rutas_exp,10)
print("Rutas de importación")
ganancias_top(rutas_imp,10)



Rutas de exportación
Las 10 rutas mas demandadas aportan el 36.6% de las ganancias.
Total de servicios realizados fue de: 2194.0 servicios

Rutas de importación
Las 10 rutas mas demandadas aportan el 52.1% de las ganancias.
Total de servicios realizados fue de: 1524.0 servicios



Como conclusión para exportaciones, se puede decir que las 10 rutas representan un 6.17% de todas las rutas, pero aprotan un 36.6% de las ganancias. De acuerdo a la regla de pareto (80-20) para realizar una recomendación, el 6.17% tendría que aportar por lo menos 24.68% de las ganancias para que estas 10 rutas sean significativas. Así que si es conveniente realizar una estrategia enfocándose en estas 10 primeras rutas para exportaciones.

Para el caso de las importaciones, las 10 rutas representan un 19.6% de todas las rutas de importación. De acuerdo a la misma regla, tendrían que aportar mínimo un 78.43% de las ganancias para que fueran representativas y lograr resultados de impacto si se decide enfocar en estas top 10 rutas. Sin embargo, se observa que solo aportan el 52% de las ganancias para importaciones. Lo que nos dice en pocas palabras, es que si se decide enfocar en este 19% de rutas, el reflejo en cuanto a las ganancias será muy pobre. Por lo que significaría mucho esfuerzo por pocas ganancias. Así que para importaciones no es tan recomendable enfocarse en el top 10 rutas, a diferencia de las exportaciones. 

## 2. Medio de transporte utilizado

In [ ]:
def analisis_transporte(direction):
    df = sldb[sldb['direction']==direction]
    df = df.groupby(['transport_mode'])
    df_top = df.count()['total_value']
    df_top = df_top.reset_index()

    ganancias = df['total_value'].sum()
    ganancias = ganancias.reset_index()

    df_top ['ganancias'] = ganancias['total_value']
    t_ganancias = df_top['ganancias'].sum()
    t_rutas = df_top['total_value'].sum()
    df_top['porcentaje_r'] = round((df_top['total_value']/t_rutas)*100,3)
    df_top['porcentaje_g'] = round((df_top['ganancias']/t_ganancias)*100,3)

    df_top = df_top.sort_values('total_value',ascending=False)

    return df_top


In [ ]:
t_exp = analisis_transporte('Exports')
t_exp

transport_mode  total_value    ganancias  porcentaje_r  porcentaje_g
3            Sea         8434  65592622000        54.738        40.954
1           Rail         2780  34505043000        18.043        21.544
2           Road         2149  27280486000        13.947        17.033
0            Air         2045  32785147000        13.272        20.470

In [ ]:
t_imp = analisis_transporte('Imports')
t_imp

transport_mode  total_value    ganancias  porcentaje_r  porcentaje_g
3            Sea         2254  34938000000        61.787        62.920
1           Rail          601   9123000000        16.475        16.430
2           Road          449   5990000000        12.308        10.787
0            Air          344   5477000000         9.430         9.863

Como conclusión se puede decir que se espera que el porcentaje de rutas de algún modo de transporte sea igual al porcentaje de ganancias que generan. Además de este indicador también se podría considerar que el modo de transporte que cuenta con menos rutas como menos ganancias podría ser reducido.

Para el caso de las exportaciones, sin lugar a duda el modo de transporte más valioso por ruta es el ferroviario. Mientras que el marítimo es el que menos valor aporta por ruta. Para poder llegar a una conclusión más exacta, podrían hacerse futuros análisis históricos para observar como han cambiado las tendencias los últimos años. Porque de ser así, aportaría información igual de importante para tomar una decisión sobre cuál transporte reducir.

Para el caso de las importaciones, el porcentaje de ganancias que aportan por porcentaje de tipos de rutas es muy parecido. Por lo que este factor parece no ser tan determinante. Por lo que se puede decir que los dos modos de transporte más importantes a considerar para ser reducidos, son tanto por caminos (roads) como los aéreos (air). Ya que estos aportan pocas ganancias y son de por si, las rutas menos utilizadas para importaciones. Sería recomendable realizar un análisis histórico como en el caso pasado, sobre todo para garantizar la calidad del servicio hacia los clientes de Synergy Logistics. 

## 3 Valor total de importaciones y exportaciones

Enfocarse en el 80% de los países que más generan valor de las exportaciones e importaciones

In [69]:
#Función que devolverá un data frame agrupado por pais de destino u origen, donde se verá un % de ganancias de cada rutas de importacion o exportacion con respecto al valor total.
#Este estará acomodado del que representa un mayor porcentaje del total, a un menor porcentaje. 

def top_valor(df,pais,top):
  imp_exp = df.groupby(pais).sum()
  total = imp_exp['suma total'].sum()
  imp_exp['porcentaje'] =  (imp_exp['suma total'] / total)*100
  imp_exp = imp_exp.sort_values('porcentaje', ascending = False)
  colcumsum = imp_exp.cumsum()['porcentaje']
  imp_exp['porcentaje acum'] = colcumsum
  imp_exp = imp_exp.reset_index()
  top = imp_exp[imp_exp['porcentaje acum'] < top]

  return top

In [75]:
ex = top_valor(rutas_exp,'origin', 100)
ex

origin   count          mean           std         min  \
0            China  1657.0  2.225607e+08  1.607859e+08  23000000.0   
1           France   919.0  1.940349e+08  1.609365e+08  29000000.0   
2              USA  1464.0  2.073022e+08  1.195892e+08  43000000.0   
3      South Korea  1134.0  6.665104e+07  4.063573e+07   7000000.0   
4           Russia   766.0  1.766204e+08  1.525876e+08  12000000.0   
5            Japan  1257.0  1.370193e+08  1.064416e+08  23000000.0   
6          Germany  1295.0  1.425838e+08  9.936885e+07  25011000.0   
7           Canada   603.0  6.905637e+07  5.944723e+07         0.0   
8            Italy   562.0  1.030107e+08  9.151343e+07  10000000.0   
9      Netherlands   796.0  2.388864e+07  2.470962e+07   8000000.0   
10           Spain   295.0  7.350371e+07  9.019050e+07         0.0   
11  United Kingdom   871.0  1.565573e+07  1.722256e+07      1000.0   
12           India   565.0  2.880562e+07  1.884579e+07   3000000.0   
13         Belgium   243.0  4.094951e+07  4.431014e+07         0.0   
14          Brazil   396.0  3.428780e+07  2.631773e+07         0.0   
15     Switzerland   249.0  3.438251e+07  2.287368e+07         0.0   
16       Australia   863.0  1.163808e+07  1.028956e+07         0.0   
17          Mexico   772.0  1.795676e+07  2.117928e+07         0.0   
18         Austria   437.0  1.594963e+04  1.717683e+04         0.0   

            25%          50%          75%          max   suma total  \
0   113252000.0  183500000.0  292250000.0  828000000.0  32977046000   
1    83502500.0  147506000.0  249509250.0  930022000.0  18614332000   
2   121501000.0  182004000.0  266751000.0  593006000.0  18355306000   
3    38000000.0   55002500.0   86502000.0  205006000.0  14621146000   
4    67500000.0  126500000.0  243750000.0  626000000.0  13223000000   
5    59003000.0  110006000.0  181759000.0  482019000.0  11948976000   
6    65001500.0  123684500.0  185298000.0  481847000.0  11343233000   
7    23500000.0   57500000.0  101750000.0  259000000.0  10684000000   
8    29261250.0   69030500.0  156767000.0  375039000.0   5366684000   
9    10007000.0   18006000.0   26011000.0  268021000.0   4120369000   
10   13750000.0   44000000.0   88000000.0  355000000.0   3418000000   
11    4003000.0    8007000.0   26003000.0   72009000.0   2777612000   
12   15250000.0   24500000.0   37000000.0   91000000.0   2626000000   
13   10500000.0   26500000.0   54500000.0  196000000.0   2588000000   
14   15000000.0   30500000.0   48750000.0  122000000.0   2274000000   
15   18250000.0   29000000.0   47000000.0  105000000.0   2154000000   
16    6000000.0    8000000.0   13750000.0   65000000.0   1650000000   
17    2159750.0   10177500.0   25306500.0   82578000.0   1419755000   
18       4000.0      12000.0      22250.0      78000.0      1155000   

    porcentaje  porcentaje acum  
0    20.589640        20.589640  
1    11.622096        32.211736  
2    11.460370        43.672105  
3     9.128899        52.801004  
4     8.255949        61.056953  
5     7.460496        68.517449  
6     7.082292        75.599741  
7     6.670692        82.270433  
8     3.350758        85.621191  
9     2.572605        88.193796  
10    2.134072        90.327868  
11    1.734238        92.062105  
12    1.639577        93.701682  
13    1.615851        95.317533  
14    1.419801        96.737334  
15    1.344877        98.082211  
16    1.030199        99.112410  
17    0.886442        99.998852  
18    0.000721        99.999573

In [76]:
imp = top_valor(rutas_imp, 'origin', 100)
imp

origin  count          mean           std        min  \
0                  China  615.0  1.375309e+08  9.690093e+07  7000000.0   
1                  Japan  445.0  1.263671e+08  8.466124e+07  7000000.0   
2                    USA  376.0  1.175110e+08  7.974145e+07  7000000.0   
3                 Mexico  366.0  7.025501e+07  4.834852e+07  3000000.0   
4                Germany  369.0  6.557100e+07  4.541197e+07  4000000.0   
5              Singapore  273.0  1.471429e+07  1.216850e+07  1000000.0   
6            South Korea  161.0  8.839712e+07  5.663430e+07  5000000.0   
7               Malaysia  199.0  3.735641e+07  3.006862e+07  3000000.0   
8                  Spain  183.0  3.773583e+07  3.220245e+07  1000000.0   
9                 France  169.0  1.573522e+07  1.174548e+07  1000000.0   
10                 Italy   86.0  3.442365e+07  2.219741e+07  3000000.0   
11             Australia  102.0  9.019608e+06  5.166496e+06  1000000.0   
12                Russia   42.0  2.026190e+07  1.229725e+07  1000000.0   
13                Canada   47.0  1.210638e+07  9.175189e+06  1000000.0   
14               Vietnam   18.0  3.000000e+07  1.838798e+07  2000000.0   
15                Brazil  119.0  1.632153e+07  8.723933e+06  1000000.0   
16  United Arab Emirates   20.0  1.855000e+07  1.165050e+07  1000000.0   

           25%          50%          75%          max   suma total  \
0   30000000.0  149000000.0  208500000.0  371000000.0  12233000000   
1   63000000.0  134000000.0  189750000.0  285000000.0   8094000000   
2   26000000.0  138000000.0  166750000.0  305000000.0   5291000000   
3   19000000.0   77000000.0  101000000.0  195000000.0   4621000000   
4   14000000.0   75000000.0   96250000.0  157000000.0   4250000000   
5    4000000.0   14000000.0   20000000.0   69000000.0   4017000000   
6   22250000.0  108500000.0  127750000.0  185000000.0   3889000000   
7   13250000.0   38500000.0   55750000.0  102000000.0   3560000000   
8    8000000.0   24500000.0   64750000.0  118000000.0   3001000000   
9    5000000.0   15000000.0   22000000.0   50000000.0   1316000000   
10  10000000.0   39000000.0   49250000.0   76000000.0   1268000000   
11   4000000.0   10500000.0   12750000.0   23000000.0    920000000   
12   4000000.0   23000000.0   29000000.0   44000000.0    851000000   
13   4000000.0   12000000.0   16500000.0   34000000.0    569000000   
14  13250000.0   33500000.0   38750000.0   57000000.0    540000000   
15  13500000.0   18500000.0   21500000.0   29000000.0    489000000   
16  11000000.0   21500000.0   25500000.0   38000000.0    371000000   

    porcentaje  porcentaje acum  
0    22.030327        22.030327  
1    14.576430        36.606757  
2     9.528526        46.135283  
3     8.321928        54.457211  
4     7.653796        62.111007  
5     7.234188        69.345195  
6     7.003674        76.348869  
7     6.411180        82.760049  
8     5.404481        88.164530  
9     2.369976        90.534505  
10    2.283533        92.818038  
11    1.656822        94.474860  
12    1.532560        96.007420  
13    1.024708        97.032128  
14    0.972482        98.004610  
15    0.880637        98.885247  
16    0.668131        99.553378

Como conclusión se puede mencionar que para el caso de las exportaciones, los primeros 7 países aportan el 75% acumulado de las ganancias. Si bien es muy buen porcentaje, 7 países representa casi la mitad de todos los países exportadores. Lo cual refleja que enfocarse en la mitad de los países, para conseguir un 75% de las ganancias, de acuerdo a la regla de pareto, no sería muy recomendable. Ya que es un gran esfuerzo el tomar la mitad de los países por un resultado no tan alto en cuestión de ganancias.

Lo mismo sucede para las importaciones. El 76.4% de las ganancias las representan un 41% de los países importadores. Es un gran esfuerzo concentrarse en tantos países, y no resulta del todo redituable. Sin embargo valdría la pena realizar un análisis de mezclas de países importadores y exportadores para saber si pueden hacerse sinergias y aprovechar al doble la oportunidad. Como por ejemplo, enfocarse en China (que aparece en el top de ambas listas) parece ser una idea muy buena. Ya que en total de ganancias, el país de china representa por lo menos un 20% de todas las ganancias, si no es que más. 

Por lo que sería aconsejable realizar esta mezcla de países para no necesariamente enfocarse en el 80% de las ganancias. Pero si en aquellos casos puntuales, como se puede observa a grandes razgos, China, USA o incluso Japón, que aparecen como grandes importadores y exportadores. 

## 4. Conclusión

Como conclusión se puede decir que cada análisis muestra una perspectiva diferente por la cual se puede obtener información muy valiosa del análisis de la gran cantidad de datos.
Se puede destacar por ejemplo, que es una buena idea enfocarse en las 10 rutas más concurridas en cuestion de importaciones, que el medio de transporte menos utilizado por Synergy Logistic es por tierra en caso de exportaciones porque aporta poco valor, mientras que para importaciones es menos importante el transporte aereo. Así como también podemos observar que solo 6 países en cuanto a rutas de importación o de exportación representan un valor de 80% de las ganancias totales de SL. Por lo que me parece de gran valor obtener información fácilmente y de una base de datos tan grande. Obtener información representativa y pertinente es vital en cualquier industria y trabajo.
